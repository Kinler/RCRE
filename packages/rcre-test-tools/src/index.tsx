export * from './RCRETestUtil';
export * from './AutomaticRobot';