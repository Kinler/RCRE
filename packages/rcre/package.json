{
    "name": "rcre",
    "version": "0.19.150",
    "main": "./dist/index.js",
    "typings": "./dist/src/index.d.ts",
    "repository": "https://github.com/andycall/RCRE",
    "homepage": "/",
    "proxy": {
        "/proxy": {
            "target": "http://localhost:8844"
        },
        "/api": {
            "target": "http://localhost:8844"
        }
    },
    "dependencies": {
        "rcre-runtime": "0.1.0",
        "axios": "^0.16.2",
        "bowser": "1.9.4",
        "classnames": "^2.2.5",
        "events": "^3.0.0",
        "lodash": "^4.17.4",
        "lru-cache": "5.1.1",
        "moment": "^2.20.1",
        "prop-types": "^15.7.2",
        "qs": "^6.5.2",
        "react": ">=16",
        "react-dom": ">=16",
        "react-redux": "^5.0.7",
        "redux": "^3.7.2"
    },
    "devDependencies": {
        "chalk": "^2.4.1"
    },
    "license": "Apache-2.0"
}
