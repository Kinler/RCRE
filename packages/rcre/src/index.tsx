/// <reference path="../global.ts" />
import react from 'react';
import reactdom from 'react-dom';
import configureStore from './data/store';
import {filter} from './core/util/filter';
import {createChild} from './core/util/createChild';
import {Store} from 'redux';
import {RootState} from './data/reducers';
import * as vm from './core/util/vm';

import './index.css';
import {containerGraph} from './core/Service/ContainerDepGraph';
import * as Container from './core/Container';
import * as DataProvider from './core/DataProvider/index';
import * as Trigger from './core/Trigger';
import * as Connect from './core/Connect';
import * as types from './types';
import * as Form from './core/Form';
import {Render} from './render';

import './core/Container/AbstractContainer';
import './core/Layout/Div/Div';
import './core/Layout/Text/Text';
import './core/Layout/Row/Row';
import {dataProviderEvent} from './core/Events/dataProviderEvent';

export * from './types';
export * from './core/util/util';
export * from './services/api';
export * from './core/Events';
export * from './core/Page';
export * from './core/Events/dataProviderEvent';
export * from './core/DataCustomer/index';
export * from './core/util/componentLoader';
export * from './services/log';
export * from './core/util/stringToPath';
export * from './language/parser/index';
export * from './data/events';

export * from './data/reducers';

export let store: Store<RootState> = configureStore();

export function clearStore() {
    store.dispatch({
        type: '_RESET_STORE_'
    });
    dataProviderEvent.clear();
    containerGraph.clear();
}

export {
    Container,
    Trigger,
    DataProvider,
    createChild,
    filter,
    containerGraph,
    vm,
    Form,
    Render,
    types,
    Connect
};

if (process.env.REMOTE_DEBUG) {
    let script = document.querySelector('script');
    let url;

    if (script) {
        url = script.src;
    }

    let msg = `you are in remote debug mode, please copy and past <script src="${url}"></script> into your webpages.`;
    console.log(msg);
}

// if (process.env.NODE_ENV !== 'test') {
//     if (isUnsupportedBrowser({
//         msie: '10',
//         chrome: '55',
//         safari: '11',
//         firefox: '10'
//     }, false, window && window.navigator.userAgent)) {
//         console.error('检测到你正在使用低版本的浏览器，请使用最新版Chrome浏览器或者Firefox浏览器');
//     }
// }
//
// if (process.env.NODE_ENV === 'test') {
//     let oldRCRE = require('@baidu/rcre-core');
//     oldRCRE.registerAddFilterCallback((name: string, fn: any) => {
//         filter.setFilter(name, fn);
//     });
//
//     // 将0.18的FuncCustomer同步到0.19
//     oldRCRE.setFucCustomerCallback((name: string, fn: any) => {
//         DataCustomer.funcCustomer.setCustomer(name, fn);
//     });
// }

export const version = __VERSION__;
export const React = react;
export const ReactDOM = reactdom;
