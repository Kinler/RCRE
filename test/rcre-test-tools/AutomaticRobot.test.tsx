import {AutomaticRobot} from 'rcre-test-tools';
import {clearStore, componentLoader, Connect} from 'rcre';
import React from 'react';

describe('AutomaticRobot', () => {
    beforeEach(() => {
        clearStore();
    });

    it('init', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                children: [
                    {
                        type: 'form',
                        name: 'test',
                        children: [
                            {
                                type: 'formItem',
                                required: true,
                                control: {
                                    type: 'input',
                                    name: 'username'
                                }
                            },
                            {
                                type: 'formItem',
                                required: true,
                                rules: [{
                                    minLength: 2
                                }],
                                control: {
                                    type: 'input',
                                    name: 'password'
                                }
                            }
                        ]
                    }
                ]
            }]
        };

        let robot = new AutomaticRobot(config);

        await robot.run([
            {
                container: 'demo',
                steps: [
                    {
                        name: 'username',
                        value: 'helloworld'
                    },
                    {
                        name: 'password',
                        value: '123'
                    }
                ]
            }
        ]);
        robot.test.unmount();
    });

    it('data not valid will cause error', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                children: [
                    {
                        type: 'form',
                        name: 'test',
                        children: [
                            {
                                type: 'formItem',
                                required: true,
                                control: {
                                    type: 'input',
                                    name: 'username'
                                }
                            },
                            {
                                type: 'formItem',
                                required: true,
                                rules: [{
                                    minLength: 2
                                }],
                                control: {
                                    type: 'input',
                                    name: 'password'
                                }
                            }
                        ]
                    }
                ]
            }]
        };

        let robot = new AutomaticRobot(config);
        try {
            await robot.run([
                {
                    container: 'demo',
                    steps: [
                        {
                            name: 'username',
                            value: 'helloworld'
                        },
                        {
                            name: 'password',
                            value: ''
                        }
                    ]
                }
            ]);
            throw new Error('robot did not throw');
        } catch (e) {
            expect(e.message).toBe('CheckState failed: password\'s form status is not valid. errmsg: 长度不能小余2');
        }

        robot.test.unmount();
    });

    it('event steps', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                dataL: {
                    username: '1'
                },
                dataCustomer: {
                    customers: [{
                        mode: 'pass',
                        name: 'testCustomer',
                        config: {
                            model: 'demo',
                            assign: {
                                passUserName: '#ES{$trigger.testCustomer.passUserName}'
                            }
                        }
                    }]
                },
                children: [{
                    type: 'form',
                    name: 'test',
                    children: [
                        {
                            type: 'formItem',
                            required: true,
                            control: {
                                type: 'input',
                                name: 'username',
                                trigger: [{
                                    event: 'onChange',
                                    targetCustomer: '$this',
                                    params: {
                                        anotherUserName: '#ES{$args.value}'
                                    }
                                }, {
                                    event: 'onChange',
                                    targetCustomer: 'testCustomer',
                                    params: {
                                        passUserName: '#ES{$args.value}'
                                    }
                                }]
                            }
                        }
                    ]
                }]
            }]
        };

        let robot = new AutomaticRobot(config);

        await robot.run([{
            container: 'demo',
            steps: [
                {
                    name: 'username',
                    value: '1234',
                    event: [{
                        eventName: 'onChange',
                        args: {
                            value: '1234'
                        }
                    }]
                }
            ]
        }]);

        expect(robot.test.getState()).toMatchSnapshot();

        robot.test.unmount();
    });

    it('Containers can still be embedded in the Steps', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                dataL: {
                    username: '1'
                },
                dataCustomer: {
                    customers: [{
                        mode: 'pass',
                        name: 'testCustomer',
                        config: {
                            model: 'demo',
                            assign: {
                                passUserName: '#ES{$trigger.testCustomer.passUserName}'
                            }
                        }
                    }]
                },
                children: [{
                    type: 'form',
                    name: 'test',
                    children: [
                        {
                            type: 'formItem',
                            required: true,
                            control: {
                                type: 'input',
                                name: 'username',
                                trigger: [{
                                    event: 'onChange',
                                    targetCustomer: '$this',
                                    params: {
                                        anotherUserName: '#ES{$args.value}'
                                    }
                                }, {
                                    event: 'onChange',
                                    targetCustomer: 'testCustomer',
                                    params: {
                                        passUserName: '#ES{$args.value}'
                                    }
                                }, {
                                    event: 'onSelect',
                                    targetCustomer: '$this',
                                    params: {}
                                }]
                            }
                        },
                        {
                            type: 'container',
                            model: 'child',
                            children: [{
                                type: 'input',
                                name: 'child-username'
                            }]
                        }
                    ]
                }]
            }]
        };

        let robot = new AutomaticRobot(config);
        await robot.run([{
            container: 'demo',
            steps: [
                {
                    name: 'username',
                    value: '1234',
                    manual: true,
                    event: [{
                        eventName: 'onChange',
                        args: {
                            value: '1234'
                        }
                    }]
                },
                {
                    container: 'child',
                    steps: [{
                        name: 'child-username',
                        value: '1234'
                    }]
                }
            ]
        }]);

        expect(robot.test.getState().container).toMatchSnapshot();

        robot.test.unmount();
    });

    it('AutoComplete NameValid', async () => {
        class NameValid extends React.PureComponent<any> {
            render() {
                return (
                    <div>this.props.value</div>
                );
            }
        }
        componentLoader.addComponent('nameValidTest', Connect.commonConnect({
            isNameValid: (value, props) => {
                if (value === '1') {
                    return true;
                }

                return false;
            }
        })(NameValid));

        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                children: [{
                    type: 'nameValidTest',
                    name: 'nameValidTest'
                }]
            }]
        };
        let robot = new AutomaticRobot(config);
        try {
            await robot.run([
                {
                    container: 'demo',
                    steps: [{
                        name: 'nameValidTest',
                        value: '2'
                    }]
                }
            ]);
        } catch (e) {
            expect(e.message).toBe('StepItem: item value is not valid. \n' +
                'Object {\n' +
                '  "name": "nameValidTest",\n' +
                '  "value": "2",\n' +
                '}');
        }

        robot.test.unmount();
    });

    it('Select NameValid', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                children: [{
                    type: 'input',
                    name: 'list'
                }]
            }]
        };

        let robot = new AutomaticRobot(config);
        try {
            await robot.run([{
                container: 'demo',
                steps: [{
                    name: 'list',
                    value: 'c'
                }]
            }]);
        } catch (e) {
            expect(e.message).toBe('StepItem: item value is not valid. \n' +
                'Object {\n' +
                '  "name": "list",\n' +
                '  "value": "c",\n' +
                '}');
        }

        robot.test.unmount();
    });

    it('Checkbox NameValid', async () => {
        let config = {
            body: [{
                type: 'container',
                model: 'demo',
                children: [{
                    type: 'checkbox',
                    name: 'template_id',
                    groups: [{key: 'BC0119', text: 'AAA', disabled: false}]
                }]
            }]
        };

        let robot = new AutomaticRobot(config);
        try {
            await robot.run([{
                container: 'demo',
                steps: [{
                    name: 'template_id',
                    value: {'AAWWE': true, 'GEFERE': true, 'BC0119': true}
                }]
            }]);
        } catch (e) {
            expect(e.message).toBe('StepItem: item value is not valid. \n' +
                'Object {\n' +
                '  "name": "template_id",\n' +
                '  "value": Object {\n' +
                '    "AAWWE": true,\n' +
                '    "BC0119": true,\n' +
                '    "GEFERE": true,\n' +
                '  },\n' +
                '}');
        }

        robot.test.unmount();
    });
});