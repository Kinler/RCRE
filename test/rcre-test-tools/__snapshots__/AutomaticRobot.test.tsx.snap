// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutomaticRobot Containers can still be embedded in the Steps 1`] = `
Object {
  "__TMP_MODEL__DO_NO_USE_IT": Object {},
  "child": Object {
    "child-username": "1234",
  },
  "demo": Object {
    "anotherUserName": "1234",
    "passUserName": "1234",
    "username": "1234",
  },
}
`;

exports[`AutomaticRobot event steps 1`] = `
Object {
  "container": Object {
    "__TMP_MODEL__DO_NO_USE_IT": Object {},
    "demo": Object {
      "anotherUserName": "1234",
      "passUserName": "1234",
      "username": "1234",
    },
  },
  "form": Object {
    "test": Object {
      "clearAfterSubmit": false,
      "control": Object {
        "nodejs": Object {
          "errorMsg": "",
          "formItemName": "nodejs",
          "formName": "test",
          "required": true,
          "rules": undefined,
          "status": "success",
          "valid": false,
        },
        "username": Object {
          "errorMsg": "",
          "formItemName": "username",
          "formName": "test",
          "required": true,
          "rules": undefined,
          "status": "success",
          "valid": true,
        },
      },
      "name": "test",
      "valid": false,
      "validateFirst": false,
    },
  },
  "trigger": Object {
    "demo": Object {
      "$SELF_PASS_CUSTOMER": Object {
        "anotherUserName": "1234",
      },
      "testCustomer": Object {
        "passUserName": "1234",
      },
    },
  },
}
`;
